uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs

```diff
@@ -24,44 +24,52 @@ import           Data.Parameterized.TraversableFC (toListFC)
 
 import qualified What4.InterpretedFloatingPoint as W4IFP
 
-import           Lang.Crucible.LLVM.MemType (MemType(..), SymType(..), FunDecl(..))
+import           Lang.Crucible.LLVM.MemType (MemType(..), SymType(..), FunDecl(..), mkStructInfo)
 
 import           UCCrux.LLVM.Errors.Panic (panic)
-import           UCCrux.LLVM.FullType.Type (FullTypeRepr(..), aliasOrFullType)
+import           UCCrux.LLVM.FullType.Type (FullTypeRepr(..), aliasOrFullType, DataLayout, crucibleDataLayout, structPackedReprToBool)
 import           UCCrux.LLVM.FullType.VarArgs (varArgsReprToBool)
 {- ORMOLU_ENABLE -}
 
 -- | Postcondition: The returned 'MemType' will not be a metadata type.
 --
 -- @toStorageType@ depends on this.
-toMemType :: FullTypeRepr m ft -> MemType
-toMemType =
+toMemType :: DataLayout m -> FullTypeRepr m ft -> MemType
+toMemType dl =
   \case
     FTIntRepr natRepr -> IntType (natValue natRepr)
     FTPtrRepr ptRepr ->
       case aliasOrFullType ptRepr of
         Left ident -> PtrType (Alias ident)
-        Right ftRepr -> PtrType (MemType (toMemType ftRepr))
+        Right ftRepr -> PtrType (MemType (toMemType dl ftRepr))
     FTFloatRepr W4IFP.SingleFloatRepr -> FloatType
     FTFloatRepr W4IFP.DoubleFloatRepr -> DoubleType
     FTFloatRepr W4IFP.X86_80FloatRepr -> X86_FP80Type
     FTFloatRepr floatInfo -> panic "toMemType" ["Illegal float type: ", show floatInfo]
     FTVoidFuncPtrRepr varArgs argsRepr ->
-      funType Nothing argsRepr (varArgsReprToBool varArgs)
+      funType dl Nothing argsRepr (varArgsReprToBool varArgs)
     FTNonVoidFuncPtrRepr varArgs retRepr argsRepr ->
-      funType (Just retRepr) argsRepr (varArgsReprToBool varArgs)
+      funType dl (Just retRepr) argsRepr (varArgsReprToBool varArgs)
     FTOpaquePtrRepr _ident -> PtrType OpaqueType
-    FTArrayRepr natRepr fullTypeRepr -> ArrayType (natValue natRepr) (toMemType fullTypeRepr)
-    FTUnboundedArrayRepr fullTypeRepr -> ArrayType 0 (toMemType fullTypeRepr)
-    FTStructRepr structInfo _ -> StructType structInfo
+    FTArrayRepr natRepr fullTypeRepr -> ArrayType (natValue natRepr) (toMemType dl fullTypeRepr)
+    FTUnboundedArrayRepr fullTypeRepr -> ArrayType 0 (toMemType dl fullTypeRepr)
+    FTStructRepr sp fields ->
+      let cdl = crucibleDataLayout dl
+          memFields = toListFC (toMemType dl) fields
+      in StructType (mkStructInfo cdl (structPackedReprToBool sp) memFields)
   where
-    funType :: Maybe (FullTypeRepr m ft) -> Ctx.Assignment (FullTypeRepr m) argTypes -> Bool -> MemType
-    funType maybeRetRepr argsRepr isVarArgs =
+    funType ::
+      DataLayout m ->
```

Comment:
Is this argument necessary? It looks like this function is only called in two places that just pass the (already visible) `dl` to this function

Reply:
It's a trade-off between using `ScopedTypeVariables` to match the `m` in this signature with the `m` in the outer signature vs just passing it in here. Happy to change it to used `ScopedTypedVariables` if you prefer it!

Reply:
ah, I had assumed scoped type variables was already enabled. I think that using scoped type variables might be less confusing in the long run, as there is no possibility of confusing values.